dataset/src/test/scala/frameless/functions/UdfTest.scala

```diff
@@ -0,0 +1,251 @@
+      val u1 = udf[X3[A, B, C], A, B, C, C](f)
+      val u2 = dataset.makeUDF(f)
+
+      val A = dataset.col[A]('a)
+      val B = dataset.col[B]('b)
+      val C = dataset.col[C]('c)
+
+      val dataset21 = dataset.select(u1(A,B,C)).collect().run().toVector
+      val dataset22 = dataset.select(u2(A,B,C)).collect().run().toVector
+      val d = data.map(x => f(x.a, x.b, x.c))
+
+      dataset21 ?= d
+      dataset22 ?= d
+    }
+
+    check(forAll(prop[Int, Int, Int]((_: Int) + (_: Int) + (_: Int)) _))
+    check(forAll(prop[String, Int, Int]((_: String).length + (_: Int) + (_: Int)) _))
+  }
+
+  test("four argument udf") {
+    def prop[A, B, C](f: (A, B, C, A) => C)(data: Vector[X3[A, B, C]])(
+      implicit
+      ea: TypedEncoder[A],
+      eb: TypedEncoder[B],
+      ec: TypedEncoder[C],
+      ex4: TypedEncoder[X3[A, B, C]],
+      ta: TypeTag[A],
+      tb: TypeTag[B],
+      tc: TypeTag[C],
+      ca: ClassTag[A],
+      cb: ClassTag[B],
+      cc: ClassTag[C]
+    ): Prop = {
+      val dataset = TypedDataset.create(data)
+      val u1 = udf[X3[A, B, C], A, B, C, A, C](f)
+      val u2 = dataset.makeUDF(f)
+
+      val A = dataset.col[A]('a)
+      val B = dataset.col[B]('b)
+      val C = dataset.col[C]('c)
+
+      val dataset21 = dataset.select(u1(A,B,C,A)).collect().run().toVector
+      val dataset22 = dataset.select(u2(A,B,C,A)).collect().run().toVector
+      val d = data.map(x => f(x.a, x.b, x.c, x.a))
+
+      dataset21 ?= d
+      dataset22 ?= d
+    }
+
+    check(forAll(prop[Int, Int, Int]((_: Int) + (_: Int) + (_: Int) + (_: Int)) _))
+    check(forAll(prop[String, Int, Int]((_: String).length + (_: Int) + (_: Int) + (_: String).length) _))
+  }
+
+  test("five argument udf") {
+    def prop[A, B, C, D, E](f: (A, B, C, D, E) => C)(data: Vector[X5[A, B, C, D, E]])(
+      implicit
+      ea: TypedEncoder[A],
+      eb: TypedEncoder[B],
+      ec: TypedEncoder[C],
+      ed: TypedEncoder[D],
+      ee: TypedEncoder[E],
+      ex4: TypedEncoder[X5[A, B, C, D, E]],
+      ta: TypeTag[A],
+      tb: TypeTag[B],
+      tc: TypeTag[C],
+      td: TypeTag[D],
+      te: TypeTag[E],
+      ca: ClassTag[A],
+      cb: ClassTag[B],
+      cc: ClassTag[C],
+      cd: ClassTag[D],
+      ce: ClassTag[E]
+    ): Prop = {
+      val dataset = TypedDataset.create(data)
+      val u1 = udf[X5[A, B, C, D, E], A, B, C, D, E, C](f)
+      val u2 = dataset.makeUDF(f)
+
+      val A = dataset.col[A]('a)
+      val B = dataset.col[B]('b)
+      val C = dataset.col[C]('c)
+      val D = dataset.col[D]('d)
+      val E = dataset.col[E]('e)
+
+      val dataset21 = dataset.select(u1(A,B,C,D,E)).collect().run().toVector
+      val dataset22 = dataset.select(u2(A,B,C,D,E)).collect().run().toVector
+      val d = data.map(x => f(x.a, x.b, x.c, x.d, x.e))
+
+      dataset21 ?= d
+      dataset22 ?= d
+    }
+
+    check(forAll(prop[Int, Int, Int, Int, Int]((_: Int) + (_: Int) + (_: Int) + (_: Int) + (_: Int)) _))
```

Comment:
I did, and it does. I am not sure why I didn't use it at the end. Let me go over this again and it might come back to me.